src/tcp.toit

```diff
@@ -5,29 +5,118 @@
 import net.tcp
 import writer
 import reader
+import tls
+import monitor
 
+import .broker
 import .transport
 import .packets
 
+
 /**
 A transport for backing an MQTT client with TCP or TLS/TCP.
+
+Supports reconnecting to the same server if constructed with the connection information.
 */
-class TcpTransport implements Transport:
-  socket_/tcp.Socket
-  writer_/writer.Writer
-  reader_/reader.BufferedReader
-
-  constructor .socket_:
-    writer_ = writer.Writer socket_
-    reader_ = reader.BufferedReader socket_
-    // Send messages immediately.
-    socket_.set_no_delay true
-
-  send packet/Packet:
-    writer_.write packet.serialize
-
-  receive --timeout/Duration?=null -> Packet?:
-    catch --unwind=(: it != DEADLINE_EXCEEDED_ERROR):
-      with_timeout timeout:
-        return Packet.deserialize reader_
-    return null
+class TcpTransport implements Transport BrokerTransport:
+  socket_ /tcp.Socket? := null
+
+  constructor socket/tcp.Socket:
+    socket_ = socket
+
+  constructor interface/tcp.Interface --host/string --port/int=1883:
+    return ReconnectingTransport_ interface --host=host --port=port
+
+  constructor.tls interface/tcp.Interface --host/string --port/int=8883
+      --root_certificates/List=[]
+      --server_name/string?=null
+      --certificate/tls.Certificate?=null:
+    return ReconnectingTlsTransport_ interface --host=host --port=port
+      --root_certificates=root_certificates
+      --server_name=server_name
+      --certificate=certificate
+
+  constructor.from_subclass_ .socket_:
+
+  write bytes/ByteArray -> int:
+    return socket_.write bytes
+
+  read -> ByteArray?:
+    return socket_.read
+
+  close:
+    if socket_: socket_.close
+    socket_ = null
+
+  is_closed -> bool:
+    return socket_ == null
+
+  supports_reconnect -> bool:
+    return false
+
+  reconnect -> none:
+    throw "UNSUPPORTED"
+
+class ReconnectingTransport_ extends TcpTransport:
+  // Reconnection information.
+  interface_ /tcp.Interface
+  host_      /string
+  port_      /int
+
+  reconnecting_mutex /monitor.Mutex := monitor.Mutex
+
+  constructor .interface_ --host/string --port/int=1883:
+    host_ = host
+    port_ = port
+    super.from_subclass_ null
+    reconnect
+
+  write bytes/ByteArray -> int:
+    return socket_.write bytes
+
+  read -> ByteArray?:
+    return socket_.read
+
+  reconnect:
+    old_socket := socket_
+    reconnecting_mutex = monitor.Mutex
+    reconnecting_mutex.do:
+      if not identical old_socket socket_: return
+      if old_socket: old_socket.close
+
+      socket := new_connection_
+      // Send messages immediately.
+      socket.set_no_delay true
```

Comment:
This will change to `socket.no_delay = true` pretty soon. Maybe you could do it already now?